### PR TITLE
sql: add mz_internal.mz_cluster

### PR DESCRIPTION
**Pull request by sploiselle**
mz_internal.mz_cluster generates the same output as `SHOW CLUSTER` but is easier to use within other SQL statements.

### Motivation

This PR adds a known-desirable feature. https://materializeinc.slack.com/archives/C02CB7L4TCG/p1666407462571889

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Review by benesch — APPROVED** [review]
LGTM but I think put it in `mz_catalog` not `mz_internal` since it seems like something users should call! (And then docs + release note.)

File: src/adapter/src/coord/dataflows.rs
Comment (mjibson) [inline src/adapter/src/coord/dataflows.rs:661]: We should figure out how this PR would work with https://github.com/MaterializeInc/materialize/issues/15195 before merging. Should this new `mz_cluster()` function always return the session setting, or should it return the cluster on which the current statement is executing? If the latter, then we would need a way to communicate that from the peek plan to here. I think it could go into `ExprPrepStyle::OneShot` which seems fine?

**Review by mjibson — APPROVED** [review]
Should we instead generalize session variable access rather than special case the cluster variable? `session_var('cluster')`?

**Comment by mjibson** [comment]
Postgres has some similar functions (https://www.postgresql.org/docs/current/functions-info.html) that all start with `current_`. Maybe this should be named `current_cluster()` instead.

**Comment by sploiselle** [comment]
@mjibson Sounds like you have more context + opinions on this design, so I'm glad to let you take this over. If you're fine just merging this in `mz_internal` so we can potentially develop the API further, let me know and I'm glad to fix the failing CI issues.

**Review by jpepin — APPROVED** [review]
Good feedback from @mjibson and @benesch here. I would personally love to get this in so we can use it internally, and iterate on the user-facing functionality as it makes sense. I will defer to the other reviewers opinions on whether the questions raised should be blocking however.

**Comment by mjibson** [comment]
I'm fine merging if this stays in `mz_internal` so it can change without backward compat guarantees.

**Comment by benesch** [comment]
There’s nothing special about mz_internal for functions right now that exempts them from backcompat, I don’t think.

**Comment by sploiselle** [comment]
@benesch My mistake--I thought `mz_internal` was understood to be unstable from users' perspectives and we could handle migrations for internal uses.

@jpepin Is the code you're running able to execute in unsafe mode or is it running on each cluster?

**Comment by mjibson** [comment]
> There’s nothing special about mz_internal for functions right now that exempts them from backcompat, I don’t think.

From https://materialize.com/docs/sql/system-catalog/:

> mz_internal, which exposes internal metadata about Materialize in an unstable format that is likely to change.

**Comment by mjibson** [comment]
Is the difference that this is a function, whereas that document describes only relations in `mz_internal`?

**Comment by mjibson** [comment]
I think this is running in cloud, so no unsafe is available.

**Comment by benesch** [comment]
> Is the difference that this is a function, whereas that document describes only relations in mz_internal?

Right. It’d be great to do the same for the functions in `mz_internal`, but I don’t think we currently do.

**Comment by benesch** [comment]
> Is the difference that this is a function, whereas that document describes only relations in mz_internal?

Right. It’d be great to do the same for the functions in `mz_internal`, but I don’t think we currently do.

**Comment by jpepin** [comment]
> @benesch My mistake--I thought `mz_internal` was understood to be unstable from users' perspectives and we could handle migrations for internal uses.
> 
> @jpepin Is the code you're running able to execute in unsafe mode or is it running on each cluster?

The use I had in mind would be running in cloud (the prometheus sql exporter), which as @mjibson noted does not have unsafe-mode afaik

File: src/adapter/src/coord/dataflows.rs
Comment (benesch) [reply src/adapter/src/coord/dataflows.rs:661]: > Should this new mz_cluster() function always return the session setting, or should it return the cluster on which the current statement is executing?

The way I want to take #15195 is to introduce a new "statement scope" for settings, which defaults to the session scope but can be overridden with the `SETTING` clause. Then there's a very natural answer here: it should return the setting in the statement scope!

**Comment by benesch** [comment]
Just in case I haven't been clear, I'm in favor of introducing a restriction in this PR to prevent creating views with calls to `mz_internal` functions, and then merging with the current name/behavior! That way we can change later if it turns out to be not quite what we want.

**Comment by sploiselle** [comment]
Took the first step toward making it possible to prohibit functions in `mz_internal` here: https://github.com/MaterializeInc/materialize/pull/15719; very much a draft

**Comment by sploiselle** [comment]
@jpepin This feature has been obviated by us implementing `current_setting(<var>)`, which gives any variable's current setting--just in case you've been waiting on this for 6 months 😂

**Pull request closed without merging**
